src/ws/cockpitauth.c

```diff
@@ -1074,6 +1074,26 @@ cockpit_session_create (CockpitAuth *self,
   return session;
 }
 
+static gboolean
+is_localhost (const char *host)
```

Comment:
this function is like a little TOC/TOU bomb waiting to go off :/

I think the constraint needs to be implemented closer to the actual place we do the connection (ie: from the ssh stuff itself).  We're also in a much better state to check it at that point.

fwiw, I just tried, on my laptop, a bunch of IPs both remote and local:

```
local 127.0.0.1 dev lo src 127.0.0.1 uid 1001 
    cache <local> 
[lis@x1 machines]$ ip route get 127.0.0.50
local 127.0.0.50 dev lo src 127.0.0.1 uid 1001 
    cache <local> 
[lis@x1 machines]$ ip route get 192.168.178.110
192.168.178.110 dev ens8u1u1u3 src 192.168.178.121 uid 1001 
    cache 
[lis@x1 machines]$ ip route get 192.168.178.109
local 192.168.178.109 dev lo src 192.168.178.109 uid 1001 
    cache <local> 

[lis@x1 machines]$ ip route get 2001:4dd1:7290:0:a6a2:b2:15dd:a454
local 2001:4dd1:7290:0:a6a2:b2:15dd:a454 from :: dev lo table local proto kernel src 2001:4dd1:7290:0:a6a2:b2:15dd:a454 metric 0 pref medium
[lis@x1 machines]$ ip route get 2001:4dd1:7290:0:a6a2:b2:15dd:a455
2001:4dd1:7290:0:a6a2:b2:15dd:a455 from :: dev ens8u1u1u3 proto ra src 2001:4dd1:7290:0:b9f3:8c92:1476:a49e metric 100 pref medium
[lis@x1 machines]$ ip route get fe80::b354:1130:c40d:dcf9
local fe80::b354:1130:c40d:dcf9 from :: dev lo table local proto kernel src fe80::b354:1130:c40d:dcf9 metric 0 pref medium
[lis@x1 machines]$ ip route get fe80::b354:1130:c40d:dcfa
fe80::b354:1130:c40d:dcfa from :: dev ens8u1u1u3 proto kernel src fe80::297f:97d5:afb:f5d metric 100 pref medium
```

so it seems that it's definitely possible to programmatically determine which IPs count as local: just use `ip route get <ip>` and see if the word `local` appears anywhere in the output.  I'm sure it would be possible to improve considerably on this starting point.

`g_inet_address_get_is_loopback()`, fwiw, is implemented like

```
  if (address->priv->family == AF_INET)
    {
      guint32 addr4 = g_ntohl (address->priv->addr.ipv4.s_addr);

      /* 127.0.0.0/8 */
      return ((addr4 & 0xff000000) == 0x7f000000);
    }
  else
    return IN6_IS_ADDR_LOOPBACK (&address->priv->addr.ipv6);
```

so it's not gonna get the IP-of-the-machine-itself case

Reply:
strace output for the above shows this interesting part:

```
socket(AF_NETLINK, SOCK_RAW|SOCK_CLOEXEC, NETLINK_ROUTE) = 3
setsockopt(3, SOL_SOCKET, SO_SNDBUF, [32768], 4) = 0
setsockopt(3, SOL_SOCKET, SO_RCVBUF, [1048576], 4) = 0
setsockopt(3, SOL_NETLINK, NETLINK_EXT_ACK, [1], 4) = 0
bind(3, {sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, 12) = 0
getsockname(3, {sa_family=AF_NETLINK, nl_pid=277435, nl_groups=00000000}, [12]) = 0
setsockopt(3, SOL_NETLINK, NETLINK_DUMP_STRICT_CHK, [1], 4) = 0
sendmsg(3, {msg_name={sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, msg_namelen=12, msg_iov=[{iov_base={{len=36, type=RTM_GETROUTE, flags=NLM_F_REQUEST, seq=1612191454, pid=0}, {rtm_family=AF_INET, rtm_dst_len=32, rtm_src_len=0, rtm_tos=0, rtm_table=RT_TABLE_UNSPEC, rtm_protocol=RTPROT_UNSPEC, rtm_scope=RT_SCOPE_UNIVERSE, rtm_type=RTN_UNSPEC, rtm_flags=RTM_F_LOOKUP_TABLE}, {{nla_len=8, nla_type=RTA_DST}, inet_addr("127.0.0.1")}}, iov_len=36}], msg_iovlen=1, msg_controllen=0, msg_flags=0}, 0) = 36
recvmsg(3, {msg_name={sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, msg_namelen=12, msg_iov=[{iov_base=NULL, iov_len=0}], msg_iovlen=1, msg_controllen=0, msg_flags=MSG_TRUNC}, MSG_PEEK|MSG_TRUNC) = 104
recvmsg(3, {msg_name={sa_family=AF_NETLINK, nl_pid=0, nl_groups=00000000}, msg_namelen=12, msg_iov=[{iov_base={{len=104, type=RTM_NEWROUTE, flags=0, seq=1612191454, pid=277435}, {rtm_family=AF_INET, rtm_dst_len=32, rtm_src_len=0, rtm_tos=0, rtm_table=RT_TABLE_MAIN, rtm_protocol=RTPROT_UNSPEC, rtm_scope=RT_SCOPE_UNIVERSE, rtm_type=RTN_LOCAL, rtm_flags=RTM_F_CLONED|0x80000000}, [{{nla_len=8, nla_type=RTA_TABLE}, RT_TABLE_MAIN}, {{nla_len=8, nla_type=RTA_DST}, inet_addr("127.0.0.1")}, {{nla_len=8, nla_type=RTA_OIF}, if_nametoindex("lo")}, {{nla_len=8, nla_type=RTA_PREFSRC}, inet_addr("127.0.0.1")}, {{nla_len=8, nla_type=RTA_UID}, 1001}, {{nla_len=36, nla_type=RTA_CACHEINFO}, {rta_clntref=2, rta_lastuse=2069731, rta_expires=0, rta_error=0, rta_used=0, rta_id=0, rta_ts=0, rta_tsage=0}}]}, iov_len=32768}], msg_iovlen=1, msg_controllen=0, msg_flags=0}, 0) = 104
```

in particular, take a look at `rtnetlink(7)` and in particular, `RTM_GETROUTE` and `RTN_LOCAL`.  There are helper macros for building these messages in `rtnetlink(3)` so it shouldn't be *so* hard...

Reply:
For what it's worth, I tested pitti's theory above, and at least if you do the simple

`iptables -A INPUT -p tcp --destination-port 22 -j DROP`

approach then `ssh localhost` will still work but `ssh my-ip-addr` will fail.  That surprised me!


Maybe that's good enough and we can skip the netlink fun.  My comments about TOC/TOU and moving the check into the connection code on the ssh side still apply though.

Reply:
btw: my TOC/TOU conerns are not theoretical.  You can give any domain you like here, so a potential attacker could feed in a name that they control, and set up their DNS server to (a) request no caching of results and (b) serve a different result the first time and the second time.

Reply:
> this function is like a little TOC/TOU bomb waiting to go off :/

To close this, we would need to do the check between name resolution and the connect syscall, right? That probably happens deep in libssh. At that point I would say this is best solved elsewhere, in the firewall config or in the sshd config.

Also, I am not sure if this PR here actually fully addresses the concerns in #15077. So I am inclined to just do nothing in Cockpit itself.

Reply:
> That probably happens deep in libssh.

For reference: ssh_connect_host_nonblocking in connect.c https://gitlab.com/libssh/libssh-mirror/-/blob/master/src/connect.c#L165

Reply:
@mvollmer : Honestly I don't really have a good gut feeling for this either. I don't really want to cripple useful functionality for some theoretical and minor issue. So fine for me to just wontfix this.

Reply:
Yep, let's do that.